File: go/validation.go
package cekat

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"
)

const (
	maxSafeInteger   int64 = 9007199254740991
	occurredAtLayout       = "2006-01-02T15:04:05.000Z"
)

// validateEvent applies the stable event rules before any delivery is attempted.
func validateEvent(eventKey string, event Event) error {
	if strings.TrimSpace(eventKey) == "" {
		return &ValidationError{Message: "event key must not be blank"}
	}
	if strings.TrimSpace(event.Email) == "" && strings.TrimSpace(event.PhoneNumber) == "" {
		return &ValidationError{Message: "event must include a non-blank email or phone number"}
	}
	if !event.OccurredAt.IsZero() {
		if year := event.OccurredAt.UTC().Year(); year < 1 || year > 9999 {
			return &ValidationError{Message: "occurred at must be between years 0001 and 9999"}
		}
	}
	return nil
}

// normalizeProperties encodes properties with encoding/json so values use their
// standard JSON representation, then verifies the result is portable JSON. The
// returned map snapshots the caller's values at call time.
func normalizeProperties(properties map[string]any) (map[string]any, error) {
	if properties == nil {
		return nil, nil
	}
	encoded, err := json.Marshal(properties)
	if err != nil {
		return nil, &ValidationError{Message: "properties is not a JSON-compatible value: " + err.Error()}
	}
	decoder := json.NewDecoder(bytes.NewReader(encoded))
	decoder.UseNumber()
	var normalized map[string]any
	if err := decoder.Decode(&normalized); err != nil {
		return nil, &ValidationError{Message: "properties is not a JSON-compatible value: " + err.Error()}
	}
	for key, value := range normalized {
		if err := validatePortableJSON(value, propertyPath("properties", key)); err != nil {
			return nil, err
		}
	}
	return normalized, nil
}

// validatePortableJSON rejects numbers that other JSON implementations cannot
// represent exactly. Decoded values are only json.Number, strings, booleans, nil,
// []any, and map[string]any.
func validatePortableJSON(value any, path string) error {
	switch typed := value.(type) {
	case json.Number:
		if !portableNumber(typed.String()) {
			return invalidJSONValue(path)
		}
	case []any:
		for index, item := range typed {
			if err := validatePortableJSON(item, fmt.Sprintf("%s[%d]", path, index)); err != nil {
				return err
			}
		}
	case map[string]any:
		for key, item := range typed {
			if err := validatePortableJSON(item, propertyPath(path, key)); err != nil {
				return err
			}
		}
	}
	return nil
}

func portableNumber(literal string) bool {
	if !strings.ContainsAny(literal, ".eE") {
		number, err := strconv.ParseInt(literal, 10, 64)
		return err == nil && number >= -maxSafeInteger && number <= maxSafeInteger
	}
	number, err := strconv.ParseFloat(literal, 64)
	if err != nil || math.IsInf(number, 0) || math.IsNaN(number) {
		return false
	}
	return number != math.Trunc(number) || math.Abs(number) <= float64(maxSafeInteger)
}

func invalidJSONValue(path string) error {
	return &ValidationError{Message: path + " is not a JSON-compatible value"}
}

func propertyPath(parent, key string) string {
	if key == "" {
		return parent + `[""]`
	}
	return parent + "." + key
}

// withOrderPaidProperties returns a copy of event whose properties include the
// required order_paid arguments. The caller's properties map is not mutated.
func withOrderPaidProperties(amount float64, currency string, event Event) (Event, error) {
	if math.IsNaN(amount) || math.IsInf(amount, 0) {
		return Event{}, &ValidationError{Message: "amount must be a finite number"}
	}
	if strings.TrimSpace(currency) == "" {
		return Event{}, &ValidationError{Message: "currency must not be blank"}
	}
	properties := make(map[string]any, len(event.Properties)+2)
	for key, value := range event.Properties {
		if key == "amount" || key == "currency" {
			return Event{}, &ValidationError{Message: "properties must not contain " + strconv.Quote(key) + "; pass it as the OrderPaid argument"}
		}
		properties[key] = value
	}
	properties["amount"] = amount
	properties["currency"] = currency
	event.Properties = properties
	return event, nil
}

func buildPayload(ctx context.Context, eventKey string, isCommon bool, event Event) (wirePayload, error) {
	if err := validateEvent(eventKey, event); err != nil {
		return wirePayload{}, err
	}

	properties, err := normalizeProperties(event.Properties)
	if err != nil {
		return wirePayload{}, err
	}
	visitorID := strings.TrimSpace(event.VisitorID)
	if visitorID == "" {
		visitorID, _ = VisitorIDFromContext(ctx)
	}
	eventID := strings.TrimSpace(event.EventID)
	if eventID == "" {
		eventID = newEventID()
	}
	occurredAt := event.OccurredAt
	if occurredAt.IsZero() {
		occurredAt = time.Now()
	}

	return wirePayload{
		EventKey:    eventKey,
		EventID:     eventID,
		OccurredAt:  occurredAt.UTC().Format(occurredAtLayout),
		ContactName: event.ContactName,
		PhoneNumber: event.PhoneNumber,
		Email:       event.Email,
		VisitorID:   visitorID,
		IsCommon:    isCommon,
		Properties:  properties,
	}, nil
}

// newEventID returns a lowercase random (version 4) UUID.
func newEventID() string {
	var id [16]byte
	if _, err := rand.Read(id[:]); err != nil {
		panic("cekat: crypto/rand failed: " + err.Error())
	}
	id[6] = (id[6] & 0x0f) | 0x40
	id[8] = (id[8] & 0x3f) | 0x80
	var text [36]byte
	hex.Encode(text[0:8], id[0:4])
	text[8] = '-'
	hex.Encode(text[9:13], id[4:6])
	text[13] = '-'
	hex.Encode(text[14:18], id[6:8])
	text[18] = '-'
	hex.Encode(text[19:23], id[8:10])
	text[23] = '-'
	hex.Encode(text[24:], id[10:])
	return string(text[:])
}

File: node/src/node/validation.ts
import { randomUUID } from 'node:crypto';

import { ValidationError } from './errors.js';
import type { EventInput, JsonValue, WirePayload } from './types.js';

export function validateAccessToken(accessToken: string): void {
  if (accessToken.trim() === '') {
    throw new ValidationError('access token must not be blank');
  }
}

/** Supplies the event ID and call time; injectable so payload tests are deterministic. */
export interface PayloadClock {
  now: () => Date;
  newEventId: () => string;
}

const systemClock: PayloadClock = { now: () => new Date(), newEventId: randomUUID };

export function buildPayload(
  eventKey: string,
  isCommon: boolean,
  event: EventInput,
  ambientVisitorId?: string,
  clock: PayloadClock = systemClock,
): WirePayload {
  validateEvent(eventKey, event);

  const explicitVisitorId = normalizeVisitorId(event.visitorId);
  const visitorId = explicitVisitorId ?? normalizeVisitorId(ambientVisitorId);
  const payload: WirePayload = {
    event_key: eventKey,
    event_id: normalizeVisitorId(event.eventId) ?? clock.newEventId(),
    occurred_at: (event.occurredAt ?? clock.now()).toISOString(),
    is_common: isCommon,
  };

  if (event.email !== undefined) payload.email = event.email;
  if (event.phoneNumber !== undefined) payload.phone_number = event.phoneNumber;
  if (event.contactName !== undefined) payload.contact_name = event.contactName;
  if (visitorId !== undefined) payload.visitor_id = visitorId;
  if (event.properties !== undefined) payload.properties = copyJsonValue(event.properties, 'properties', new WeakSet<object>()) as Record<string, JsonValue>;

  return payload;
}

/**
 * Returns a copy of `event` whose properties include the required order_paid arguments.
 * Non-object events are returned unchanged so ordinary event validation reports them.
 */
export function withOrderPaidProperties(amount: number, currency: string, event: EventInput): EventInput {
  if (typeof amount !== 'number' || !Number.isFinite(amount)) {
    throw new ValidationError('amount must be a finite number');
  }
  if (typeof currency !== 'string' || currency.trim() === '') {
    throw new ValidationError('currency must not be blank');
  }
  if (typeof event !== 'object' || event === null) return event;
  const properties = event.properties;
  if (properties !== undefined && properties !== null && typeof properties === 'object') {
    for (const key of ['amount', 'currency']) {
      if (Object.hasOwn(properties, key)) {
        throw new ValidationError(`properties must not contain "${key}"; pass it as the orderPaid argument`);
      }
    }
  }
  return { ...event, properties: { ...properties, amount, currency } };
}

function validateEvent(eventKey: string, event: EventInput): void {
  if (typeof eventKey !== 'string' || eventKey.trim() === '') {
    throw new ValidationError('event key must not be blank');
  }
  if (typeof event !== 'object' || event === null) {
    throw new ValidationError('event must be an object');
  }
  for (const field of ['email', 'phoneNumber', 'contactName', 'visitorId', 'eventId'] as const) {
    if (event[field] !== undefined && typeof event[field] !== 'string') {
      throw new ValidationError(`${field} must be a string`);
    }
  }
  if (event.occurredAt !== undefined) {
    const year = event.occurredAt instanceof Date ? event.occurredAt.getUTCFullYear() : Number.NaN;
    if (!(year >= 1 && year <= 9999)) {
      throw new ValidationError('occurredAt must be a valid Date between years 0001 and 9999');
    }
  }
  if (event.email?.trim() === '' || event.email === undefined) {
    if (event.phoneNumber?.trim() === '' || event.phoneNumber === undefined) {
      throw new ValidationError('event must include a non-blank email or phone number');
    }
  }
}

function normalizeVisitorId(visitorId: string | undefined): string | undefined {
  if (visitorId === undefined) return undefined;
  const normalized = visitorId.trim();
  return normalized === '' ? undefined : normalized;
}

function copyJsonValue(value: unknown, path: string, ancestors: WeakSet<object>): JsonValue {
  if (value === null || typeof value === 'boolean' || typeof value === 'string') return value;

  if (typeof value === 'number') {
    if (!Number.isFinite(value) || (Number.isInteger(value) && !Number.isSafeInteger(value))) {
      invalidValue(path);
    }
    return value;
  }

  if (typeof value !== 'object') invalidValue(path);

  const objectValue = value as object;
  if (ancestors.has(objectValue)) {
    throw new ValidationError(`${path} contains a cycle`);
  }
  ancestors.add(objectValue);
  try {
    if (Object.getOwnPropertySymbols(objectValue).some((key) => Object.prototype.propertyIsEnumerable.call(objectValue, key))) {
      invalidValue(path);
    }

    if (Array.isArray(objectValue)) {
      for (const key of Object.keys(objectValue)) {
        if (!isArrayIndex(key, objectValue.length)) invalidValue(path);
      }
      const copied: JsonValue[] = [];
      for (let index = 0; index < objectValue.length; index += 1) {
        const key = String(index);
        const descriptor = Object.getOwnPropertyDescriptor(objectValue, key);
        if (descriptor === undefined || !descriptor.enumerable || !Object.hasOwn(descriptor, 'value')) {
          invalidValue(`${path}[${index}]`);
        }
        copied.push(copyJsonValue(descriptor.value, `${path}[${index}]`, ancestors));
      }
      return copied;
    }

    const prototype = Object.getPrototypeOf(objectValue);
    if (prototype !== Object.prototype && prototype !== null) invalidValue(path);

    const copied: { [key: string]: JsonValue } = {};
    for (const key of Object.keys(objectValue)) {
      const descriptor = Object.getOwnPropertyDescriptor(objectValue, key);
      if (descriptor === undefined || !Object.hasOwn(descriptor, 'value')) {
        invalidValue(propertyPath(path, key));
      }
      Object.defineProperty(copied, key, {
        configurable: true,
        enumerable: true,
        writable: true,
        value: copyJsonValue(descriptor.value, propertyPath(path, key), ancestors),
      });
    }
    return copied;
  } finally {
    ancestors.delete(objectValue);
  }
}

function isArrayIndex(key: string, length: number): boolean {
  const index = Number(key);
  return Number.isSafeInteger(index) && index >= 0 && index < length && String(index) === key;
}

function invalidValue(path: string): never {
  throw new ValidationError(`${path} is not a JSON-compatible value`);
}

function propertyPath(parent: string, key: string): string {
  return key === '' ? `${parent}[""]` : `${parent}.${key}`;
}
